File: simplypdf/src/main/java/com/wwdablu/soumya/simplypdf/jsonengine/TableConverter.java
package com.wwdablu.soumya.simplypdf.jsonengine;

import android.text.TextUtils;

import com.google.gson.Gson;
import com.wwdablu.soumya.simplypdf.composers.Composer;
import com.wwdablu.soumya.simplypdf.composers.TableComposer;
import com.wwdablu.soumya.simplypdf.composers.models.TableProperties;
import com.wwdablu.soumya.simplypdf.composers.models.TextProperties;
import com.wwdablu.soumya.simplypdf.composers.models.cell.Cell;
import com.wwdablu.soumya.simplypdf.composers.models.cell.TextCell;

import org.json.JSONArray;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

class TableConverter extends BaseConverter {

    @Override
    protected void generate(Composer composer, JSONObject compose) throws Exception {

        if(!(composer instanceof TableComposer)) {
            return;
        }

        TableComposer tableComposer = (TableComposer) composer;

        String tableProperties = getProperties(compose);
        tableComposer.setProperties(TextUtils.isEmpty(tableProperties) ? null :
                new Gson().fromJson(tableProperties, TableProperties.class));

        //Generate the cell information
        List<List<Cell>> rowList = new ArrayList<>();

        JSONArray rowArray = compose.getJSONArray(Node.COMPOSER_TABLE_CONTENTS);
        int rowCount = rowArray.length();
        for(int rowIndex = 0; rowIndex < rowCount; rowIndex++) {

            JSONObject rowObject = rowArray.getJSONObject(rowIndex);
            JSONArray columnArray = rowObject.getJSONArray(Node.COMPOSER_TABLE_ROW);

            int colCount = columnArray.length();
            ArrayList<Cell> columnList = new ArrayList<>(colCount);
            rowList.add(columnList);
            for(int colIndex = 0; colIndex < colCount; colIndex++) {

                JSONObject colObject = columnArray.getJSONObject(colIndex);

                switch (colObject.getString(Node.TYPE).toLowerCase()) {

                    case Node.TYPE_TEXT:
                        String colTextProperties = getProperties(colObject);
                        TextCell textCell = new TextCell(
                                colObject.getString(Node.COMPOSER_TEXT_CONTENT),
                                TextUtils.isEmpty(colTextProperties) ? null :
                                        new Gson().fromJson(colTextProperties, TextProperties.class),
                                tableComposer.resolveCellWidth(colObject.getInt(Node.COMPOSER_TABLE_WIDTH))
                        );
                        columnList.add(textCell);
                        break;
                }
            }
        }

        tableComposer.draw(rowList);
    }
}

File: simplypdf/src/main/java/com/wwdablu/soumya/simplypdf/jsonengine/TableConverter.kt
package com.wwdablu.soumya.simplypdf.jsonengine

import android.text.TextUtils
import com.google.gson.Gson
import com.wwdablu.soumya.simplypdf.SimplyPdfDocument
import com.wwdablu.soumya.simplypdf.composers.TableComposer
import com.wwdablu.soumya.simplypdf.composers.properties.TableProperties
import com.wwdablu.soumya.simplypdf.composers.properties.TextProperties
import com.wwdablu.soumya.simplypdf.composers.properties.cell.Cell
import com.wwdablu.soumya.simplypdf.composers.properties.cell.ImageCell
import com.wwdablu.soumya.simplypdf.composers.properties.cell.TextCell
import com.wwdablu.soumya.simplypdf.jsonengine.base.ComposerConverter
import org.json.JSONObject
import java.util.*

internal class TableConverter(simplyPdfDocument: SimplyPdfDocument) : ComposerConverter(simplyPdfDocument) {

    @Throws(Exception::class)
    override fun generate(composeJsonObject: JSONObject) {

        val tableProperties = getProperties(composeJsonObject)
        val composer = simplyPdfDocument.table

        //Generate the cell information
        val rowList: MutableList<List<Cell>> = ArrayList()
        val rowArray = composeJsonObject.getJSONArray(Node.COMPOSER_TABLE_CONTENTS)
        val rowCount = rowArray.length()
        for (rowIndex in 0 until rowCount) {
            val rowObject = rowArray.getJSONObject(rowIndex)
            val columnArray = rowObject.getJSONArray(Node.COMPOSER_TABLE_ROW)
            val colCount = columnArray.length()
            val columnList = ArrayList<Cell>(colCount)
            rowList.add(columnList)
            for (colIndex in 0 until colCount) {
                val colObject = columnArray.getJSONObject(colIndex)
                columnList.add(cellConverter(colObject, composer))
            }
        }
        composer.draw(rowList, if (TextUtils.isEmpty(tableProperties)) TableProperties() else Gson().fromJson(
            tableProperties, TableProperties::class.java))
    }

    internal fun cellConverter(jsonObject: JSONObject, composer: TableComposer) : Cell {

        val widthPercent = if(jsonObject.has(Node.COMPOSER_TABLE_WIDTH)) {
            jsonObject.getInt(Node.COMPOSER_TABLE_WIDTH)
        } else {
            100
        }

        return when (jsonObject.getString(Node.TYPE).lowercase()) {
            Node.TYPE_TEXT -> {
                val colTextProperties = getProperties(jsonObject)
                TextCell(
                    jsonObject.getString(Node.COMPOSER_TEXT_CONTENT),
                    if (TextUtils.isEmpty(colTextProperties)) TextProperties() else Gson().fromJson(
                        colTextProperties,
                        TextProperties::class.java
                    ),
                    composer.resolveCellWidth(widthPercent)
                )
            }
            else -> {
                TextCell("", TextProperties().apply { textSize = 0 },
                    composer.resolveCellWidth(widthPercent)
                )
            }
        }
    }

    override fun getTypeHandler(): String = "table"
}